shadowsocks/stream.go

```diff
@@ -100,12 +155,81 @@ func isZero(b []byte) bool {
 	return true
 }
 
+// Returns the slices of sw.buf in which to place plaintext for encryption.
+func (sw *shadowsocksWriter) buffers() (sizeBuf, payloadBuf []byte) {
+	// sw.buf starts with the salt.
+	saltSize := sw.ssCipher.SaltSize()
+
+	// Each Shadowsocks-TCP message consists of a fixed-length size block,
+	// followed by a variable-length payload block.
+	sizeBuf = sw.buf[saltSize : saltSize+2]
+	payloadStart := saltSize + 2 + sw.aead.Overhead()
+	payloadBuf = sw.buf[payloadStart : payloadStart+payloadSizeMask]
+	return
+}
+
 func (sw *shadowsocksWriter) ReadFrom(r io.Reader) (int64, error) {
 	if err := sw.init(); err != nil {
 		return 0, err
 	}
 	var written int64
+	var err error
+	_, payloadBuf := sw.buffers()
 
+	// Special case: one thread-safe read, if necessary
+	sw.mu.Lock()
+	if sw.needFlush {
+		pending := sw.pending
+
+		sw.mu.Unlock()
+		// It's not safe to read into payloadBuf, which may be
+		// modified on the flush thread.
+		// Any size is acceptable here, but this is the largest value
+		// for which a single call to enqueue() is sufficient.
+		readBuf := make([]byte, len(payloadBuf)-pending)
```

Comment:
You don't need this allocation.
You can use `sw.buf[saltSize + 2 + sw.aead,Overhead() + sw.pending + sw.aead.Overhead():]`

The `enqueue` call later will shift things to the right position.

Reply:
Beautiful.  Done.